### PR TITLE
[FEATURE] `TableAsset.test_connection()` should fail if table is not queryable.

### PR DESCRIPTION
**Pull request by Kilo59**
This PR changes `TableAsset.test_connection()` to attempt to query the `table_name` in question rather than using the `sqlalchemy` `Inspector`. 
Without this, a user could successfully add a `TableAsset` whose table exists and passes the `.test_connection()` check, but it isn't actually usable because the user/datasource config cannot query the table.

## TODO

- [x] Add tests to ensure unquerable assets raise `TestConnectionError` when adding a `TableAsset` to a `SnowflakeDatasource`
- [x] Create and add snowflake CI role for additional testing

### CODE REVIEW

File: tests/datasource/fluent/integration/test_connections.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            param(
+                "snowflake://ci:${SNOWFLAKE_CI_USER_PASSWORD}@${SNOWFLAKE_CI_ACCOUNT}?warehouse=ci&role=ci",
+                id="missing database + schema",
+            ),
+            param(
+                "snowflake://ci:${SNOWFLAKE_CI_USER_PASSWORD}@${SNOWFLAKE_CI_ACCOUNT}/ci?warehouse=ci&role=ci",
+                id="missing schema",
+            ),
+        ],
+    )
```
Comment (Kilo59) [inline tests/datasource/fluent/integration/test_connections.py]: @anthonyburdi I haven't actually changed the implementation of `TableAsset.test_connection()` yet so I was expecting these all to fail.
However, none of them do.

I'm using the sqlalchemy inspector to select a `table`, then I ensure that table cannot be queried (it raises a sqlalchemy programming error), but in each of the cases, the `.test_connection()` actually does fail.

Do you have any suggestions for how to reproduce the issue of the asset being un-selectable but `test_connection()` passing?

Comment (anthonyburdi) [reply tests/datasource/fluent/integration/test_connections.py]: I'm guessing it is because the `ci` user has [select](https://docs.snowflake.com/en/user-guide/security-access-control-privileges#table-privileges) permissions. How about creating another user `ci_without_select_permissions` and using that in this test? I think the specific issue we (me and @josectobar) were seeing was more about the select (or [usage](https://docs.snowflake.com/en/user-guide/security-access-control-privileges#schema-privileges)) permissions rather than the connection string parameters.

File: tests/datasource/fluent/integration/test_connections.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    @pytest.mark.parametrize(
+        "connection_string",
+        [
+            param(
+                "snowflake://ci:${SNOWFLAKE_CI_USER_PASSWORD}@${SNOWFLAKE_CI_ACCOUNT}/ci/public?warehouse=ci",
+                id="missing role",
+            ),
+            param(
+                "snowflake://ci:${SNOWFLAKE_CI_USER_PASSWORD}@${SNOWFLAKE_CI_ACCOUNT}/ci/public?warehouse=ci&role=no_select",
+                id="role wo select",
```
Comment (Kilo59) [inline tests/datasource/fluent/integration/test_connections.py]: @allensallinger to create this role.

File: tests/datasource/fluent/integration/test_connections.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        "connection_string",
+        [
+            param(
+                "snowflake://ci:${SNOWFLAKE_CI_USER_PASSWORD}@${SNOWFLAKE_CI_ACCOUNT}/ci/public?warehouse=ci",
+                id="missing role",
+            ),
+            param(
+                "snowflake://ci:${SNOWFLAKE_CI_USER_PASSWORD}@${SNOWFLAKE_CI_ACCOUNT}/ci/public?warehouse=ci&role=ci_no_select",
+                id="role wo select",
+            ),
```
Comment (Kilo59) [inline tests/datasource/fluent/integration/test_connections.py:33]: @allensallinger 
```
Attempt to connect to datasource failed with the following error message:
(snowflake.connector.errors.DatabaseError) 250001 (08001):
None: Failed to connect to DB: oca29081.us-east-1.snowflakecomputing.com:443. 
Role 'CI_NO_SELECT' specified in the connect string is not granted to this user. 
Contact your local system administrator, or attempt to login with another role, e.g. PUBLIC.
```

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
What if they expect the table to be empty (I'm thinking maybe this is a setup step at the beginning of a pipeline)? I think this logic should only apply to `QueryAsset`s, not `TableAsset`s.

You can use some expectations on an empty table (expect row count to be zero, expect columns to be list, expect types to be in list, etc).

File: great_expectations/datasource/fluent/sql_datasource.py [hunk 4]
```diff
@@ -921,14 +921,19 @@ def test_connection(self) -> None:
...
-            table_name=self.table_name,
-            schema=self.schema_name,
-        ):
-            raise TestConnectionError(
-                f"Attempt to connect to table: {self.qualified_name} failed because the table"
-                f" {self.table_name} does not exist."
+        try:
+            with engine.connect() as connection:
+                connection.execute(
+                    sa.text(f"SELECT * FROM {self.qualified_name} LIMIT 1;")
```
Comment (allensallinger) [inline great_expectations/datasource/fluent/sql_datasource.py]: Will we be able to perform a select here since a datasource doesn't know about a given table like a data asset would?

**Comment by Kilo59** [comment]
@NathanFarmer an empty table will pass this check. We aren't actually checking the result, simply that it executes without error.

I could update this query so that it's more obvious like...
```sql
SELECT 1 from foobar LIMIT 1;
```

File: great_expectations/datasource/fluent/sql_datasource.py (hunk 4 above)
Comment (Kilo59) [reply great_expectations/datasource/fluent/sql_datasource.py]: This is the `TableAsset.test_connection()` method.

Comment (NathanFarmer) [reply great_expectations/datasource/fluent/sql_datasource.py]: Can we use `sa.select` here?

File: tests/datasource/fluent/integration/test_connections.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+        # the table is queryable so the `add_table_asset()` should pass the test_connection step
+        asset = snowflake_ds.add_table_asset(
+            name="reachable asset", table_name=table_name
+        )
+        print(f"\n  Yay, asset was created!\n{asset!r}")
+
+
+if __name__ == "__main__":
+    pytest.main()
```
Comment (NathanFarmer) [inline tests/datasource/fluent/integration/test_connections.py]: Why is this needed?

**Review by NathanFarmer — APPROVED** [review]

File: tests/datasource/fluent/integration/test_connections.py (hunk 5 above)
Comment (Kilo59) [reply tests/datasource/fluent/integration/test_connections.py]: It isn't needed but it makes it easier to run and debug a particular test module from an IDE.
It should be `pytest.min([__file__, "-vv"])`.

So if I hit the shortcut for running the current file the IDE will usually pass the current file to `python <FILE_NAME>`, which won't actually run `pytest`, but with this `if __name__ == "__main__"` it will. 

Also, as a side benefit, it helps anchor the end of the file in the git diff.

File: great_expectations/datasource/fluent/sql_datasource.py (hunk 4 above)
Comment (Kilo59) [reply great_expectations/datasource/fluent/sql_datasource.py]: Done

File: great_expectations/datasource/fluent/sql_datasource.py [hunk 6]
```diff
@@ -921,14 +921,19 @@ def test_connection(self) -> None:
...
-            table_name=self.table_name,
-            schema=self.schema_name,
-        ):
-            raise TestConnectionError(
-                f"Attempt to connect to table: {self.qualified_name} failed because the table"
-                f" {self.table_name} does not exist."
+        try:
+            with engine.connect() as connection:
+                table = sa.table(self.table_name, schema=self.schema_name)
+                # don't need to fetch any data, just want to make sure the table is accessible
```
Comment (anthonyburdi) [inline great_expectations/datasource/fluent/sql_datasource.py:927]: Nice

File: great_expectations/datasource/fluent/sql_datasource.py [hunk 7]
```diff
@@ -921,14 +921,19 @@ def test_connection(self) -> None:
...
+                connection.execute(sa.select(1, table).limit(1))
+        except Exception as query_error:
+            LOGGER.info(
+                f"{self.name} `.test_connection()` query failed: {query_error!r}"
             )
+            raise TestConnectionError(
+                f"Attempt to connect to table: {self.qualified_name} failed because the test query "
+                f"failed. Ensure the table exists and the user has access: {query_error}"
```
Comment (anthonyburdi) [inline great_expectations/datasource/fluent/sql_datasource.py]: What about something like this?
```suggestion
                f"Attempt to connect to table: {self.qualified_name} failed because the test query "
                f"failed. Ensure the table exists and the user has access to select data from the table: {query_error}"
```

**Review by anthonyburdi — APPROVED** [review]
Nice!! Just an optional suggestion on wording.
